navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt

```diff
@@ -0,0 +1,129 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.rememberSaveableStateHolder
+import androidx.lifecycle.Lifecycle
+import androidx.navigation.FloatingWindow
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.NavDestination
+import androidx.navigation.NavOptions
+import androidx.navigation.Navigator
+import androidx.navigation.NavigatorState
+import androidx.navigation.compose.material.BottomSheetNavigator.Destination
+
+/**
+ * Create and remember a [BottomSheetNavigator]
+ *
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+public fun rememberBottomSheetNavigator(
+    sheetState: ModalBottomSheetState =
+        rememberModalBottomSheetState(ModalBottomSheetValue.Hidden)
+): BottomSheetNavigator = remember(sheetState) {
+    BottomSheetNavigator(sheetState = sheetState)
+}
+
+/**
+ * Navigator that drives a [ModalBottomSheetState] for use of [ModalBottomSheetLayout]s
+ * with the navigation library. Every destination using this Navigator must set a valid
+ * [Composable] by setting it directly on an instantiated [Destination] or calling
+ * [androidx.navigation.compose.material.bottomSheet].
+ *
+ * <b>The [sheetContent] [Composable] will always host the latest entry of the back stack. When
+ * navigating from a [BottomSheetNavigator.Destination] to another
+ * [BottomSheetNavigator.Destination], the content of the sheet will be replaced instead of a
+ * new bottom sheet being shown.</b>
+ *
+ * When the sheet is dismissed by the user, the [state]'s [NavigatorState.backStack] will be popped.
+ *
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Navigator.Name("BottomSheetNavigator")
+public class BottomSheetNavigator(
+    val sheetState: ModalBottomSheetState
+) : Navigator<BottomSheetNavigator.Destination>() {
+
+    /**
+     * A [Composable] function that hosts the current sheet content. This should be set as
+     * sheetContent of your [ModalBottomSheetLayout].
+     */
+    public val sheetContent: @Composable ColumnScope.() -> Unit = @Composable {
+        val columnScope = this
+        val saveableStateHolder = rememberSaveableStateHolder()
+        val backStackEntries by state.backStack.collectAsState()
+        val transitionsInProgress by state.transitionsInProgress.collectAsState()
+
+        // We always replace the sheet's content instead of overlaying and nesting floating
+        // window destinations. That means that only *one* concurrent destination is supported by
+        // this navigator.
+        val latestEntry = backStackEntries.lastOrNull { entry ->
```

Comment:
So when the system back button is pressed and `NavController` calls `popBackStack()` on the Navigator, how does this Navigator handle that? The default behavior of `popBackStack()` will call `pop`, which will immediately destroy the state of the `NavBackStackEntry` and remove it from `backStackEntries`. Should you be calling `popWithTransition()` there and looking at `transitionsInProgress`.lastOrNull()` and `backStackEntries` here to know what to be hiding?

Reply:
So when the entry gets removed from `backStackEntries`, the `SheetContentHost` will be recomposed with the new entry or `null`, indicating that the sheet should be empty. `SheetContentHost` has a `DisposableEffect` keyed to its `backStackEntry` to hide the sheet whenever this sheet content leaves the composition - so effectively, the sheet is hidden after the entry has been destroyed. Writing that, yeah sounds like we should use `popWithTransition` for `popBackStack`.

I thought about

```kotlin
override fun popBackStack(popUpTo: NavBackStackEntry, savedState: Boolean) {
    state.popWithTransition(popUpTo, savedState)
}
```

```kotlin
SheetContentHost(
   ...,
   onSheetDismissed = { backStackEntry ->
      transitionsInProgress[backStackEntry]?.onTransitionComplete() ?: state.pop(backStackEntry, saveState = false)
   }
)
```

That'd solve the issue of always popping with a transition and give us the possibility to fall back on `pop`, i.e. when the sheet is dismissed by a user gesture.

Reply:
That seems reasonable to me, yeah.